File: backend/app/services/union_find.py
"""
Union-Find (Disjoint Set Union) data structure with path compression and union by rank.
Used for detecting and managing conflicts in PCB routing.
"""
class UnionFind:
    def __init__(self, size(self, size: int)"""
        self.parent = list(range(size))
        self.rank = [0] * size

    def find(self, x):
        """Find the root of element x with path compression."""
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        """Union the sets containing x and y using union by rank."""
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x == root_y:
            return False  # Already in the same set

        # Union by rank
        if self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
        elif self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += 1
        return True

    def connected(self, x, y):
        """Check if x and y are in the same set."""
        return self.find(x) == self.find(y)

File: frontend/src/lib/unionFind.ts
/**
 * Disjoint-Set Union-Find data structure with Path Compression and Union by Rank.
 * Used in Rip-Up-and-Reroute to cluster mutually conflicting nets for diagnostics.
 */
export class UnionFind {
  private parent: Map<string, string>;
  private rank: Map<string, number>;

  constructor(elements: string[]) {
    this.parent = new Map();
    this.rank = new Map();
    for (const elem of elements) {
      this.parent.set(elem, elem);
      this.rank.set(elem, 0);
    }
  }

  public find(item: string): string {
    if (!this.parent.has(item)) {
      this.parent.set(item, item);
      this.rank.set(item, 0);
      return item;
    }
    const root = this.parent.get(item)!;
    if (root !== item) {
      // Path compression
      const compressedRoot = this.find(root);
      this.parent.set(item, compressedRoot);
      return compressedRoot;
    }
    return item;
  }

  public union(x: string, y: string): boolean {
    const rootX = this.find(x);
    const rootY = this.find(y);

    if (rootX === rootY) return false;

    const rankX = this.rank.get(rootX) || 0;
    const rankY = this.rank.get(rootY) || 0;

    // Union by rank
    if (rankX < rankY) {
      this.parent.set(rootX, rootY);
    } else if (rankX > rankY) {
      this.parent.set(rootY, rootX);
    } else {
      this.parent.set(rootY, rootX);
      this.rank.set(rootX, rankX + 1);
    }
    return true;
  }

  public getClusters(): Array<{ clusterId: number; netIds: string[] }> {
    const clustersMap = new Map<string, string[]>();
    for (const key of this.parent.keys()) {
      const root = this.find(key);
      if (!clustersMap.has(root)) {
        clustersMap.set(root, []);
      }
      clustersMap.get(root)!.push(key);
    }

    const result: Array<{ clusterId: number; netIds: string[] }> = [];
    let id = 1;
    for (const netIds of clustersMap.values()) {
      if (netIds.length > 1) {
        result.push({ clusterId: id++, netIds });
      }
    }
    return result;
  }
}
